File: app/src/main/java/com/infinite/narrative/engine/NarrativeEngine.kt
package com.infinite.narrative.engine

import com.infinite.narrative.ai.AIClient
import com.infinite.narrative.ai.ContentFilter
import com.infinite.narrative.ai.model.*
import com.infinite.narrative.data.model.*
import kotlinx.coroutines.flow.Flow
import com.infinite.narrative.ai.model.NarrativeContext


/**
 * 动态叙事引擎 - 核心的AI生成系统
 */
class NarrativeEngine(
    private val aiClient: AIClient,
    private val contentFilter: ContentFilter,
    private val worldStateManager: WorldStateManager,
    private val contextManager: NarrativeContextManager
) {

    /**
     * 生成新的故事片段
     * @param playerId 玩家ID
     * @param worldId 世界ID
     * @param playerChoice 玩家选择的选项ID（可选）
     * @return 生成的故事响应
     */
    suspend fun generateStory(
        playerId: String,
        worldId: String,
        playerChoice: String? = null
    ): StoryGenerationResult {

        // 1. 构建叙事上下文
        val context = contextManager.buildNarrativeContext(
            playerId = playerId,
            worldId = worldId,
            playerChoice = playerChoice
        )

        // 2. 生成AI响应
        val rawResponse = aiClient.generateNarrative(context)

        // 3. 内容安全过滤
        val filteredResponse = filterResponse(rawResponse)

        // 4. 更新世界状态
        val updatedState = worldStateManager.updateWorldState(
            playerId = playerId,
            worldId = worldId,
            response = filteredResponse,
            context = context
        )

        // 5. 保存上下文快照
        contextManager.saveContextSnapshot(
            playerId = playerId,
            worldId = worldId,
            context = context,
            response = filteredResponse
        )

        return StoryGenerationResult(
            storySegment = filteredResponse.storySegment,
            options = filteredResponse.generatedOptions,
            worldState = updatedState,
            confidenceScore = filteredResponse.confidenceScore
        )
    }

    /**
     * 构建世界选择上下文
     * @param playerId 玩家ID
     * @return 世界选择建议
     */
    suspend fun generateWorldSelection(playerId: String): WorldSelectionResult {
        val playerProgress = worldStateManager.getPlayerProgress(playerId)
        val availableWorlds = worldStateManager.getAvailableWorlds(playerId)

        // 构建世界选择上下文
        val context = NarrativeContext(
            systemInstruction = "你是一位世界引导者，需要根据玩家的成长和经历，推荐最适合的下一个故事世界。每个世界都有独特的风格和挑战。",
            worldState = WorldState(
                currentWorld = "nexus", // 叙事行者的大本营
                keyCharacters = emptyMap(),
                unlockedLocations = listOf("万象之根", "叙事长廊"),
                corePuzzles = emptyMap(),
                storyThreads = emptyList()
            ),
            recentStory = buildPlayerHistorySummary(playerProgress),
            playerAttributes = playerProgress.playerAttributes,
            availableWorlds = availableWorlds,
            activeQuests = playerProgress.activeQuests,
            inventoryItems = playerProgress.inventoryItems,
            factionRelations = playerProgress.factionRelations
        )

        val response = aiClient.generateNarrative(context)
        val filteredResponse = filterResponse(response)

        return WorldSelectionResult(
            recommendation = filteredResponse.storySegment,
            worldOptions = extractWorldOptions(filteredResponse.generatedOptions),
            reasoning = extractWorldReasoning(filteredResponse.storySegment)
        )
    }

    /**
     * 过滤AI生成的响应
     */
    private fun filterResponse(response: NarrativeResponse): NarrativeResponse {
        val filteredStory = contentFilter.filterContent(response.storySegment)

        val filteredOptions = response.generatedOptions.map { option ->
            val filteredText = contentFilter.filterContent(option.text)
            val filteredDescription = contentFilter.filterContent(option.description)

            option.copy(
                text = filteredText.filtered,
                description = filteredDescription.filtered
            )
        }

        return response.copy(
            storySegment = filteredStory.filtered,
            generatedOptions = filteredOptions
        )
    }

    /**
     * 构建玩家历史摘要
     */
    private fun buildPlayerHistorySummary(progress: PlayerProgress): String {
        val sb = StringBuilder()
        sb.append("玩家叙事行者的历史摘要：\n")

        if (progress.completedWorlds.isNotEmpty()) {
            sb.append("已完成的世界：${progress.completedWorlds.joinToString(", ")}\n")
        }

        if (progress.acquiredItems.isNotEmpty()) {
            sb.append("获得的独特物品：${progress.acquiredItems.take(5).joinToString(", ")}\n")
        }

        if (progress.keyDecisions.isNotEmpty()) {
            sb.append("关键抉择：${progress.keyDecisions.take(3).joinToString("； ")}\n")
        }

        sb.append("当前属性：洞察力${progress.playerAttributes.insight}，说服力${progress.playerAttributes.persuasion}，魄力${progress.playerAttributes.魄力}")

        return sb.toString()
    }

    /**
     * 从选项中提取世界选择
     */
    private fun extractWorldOptions(options: List<StoryOption>): List<WorldOption> {
        return options.map { option ->
            WorldOption(
                worldId = option.optionId,
                name = option.text,
                description = option.description,
                requiredAttributes = option.requiredAttributes ?: emptyMap()
            )
        }
    }

    /**
     * 从故事中提取世界推荐理由
     */
    private fun extractWorldReasoning(story: String): String {
        // 简单实现：提取包含"因为"、"所以"、"适合"等关键词的句子
        return story.lines()
            .filter { line ->
                line.contains("因为") || line.contains("所以") ||
                        line.contains("适合") || line.contains("推荐")
            }
            .joinToString(" ")
            .ifEmpty { "基于你的经历和能力，这个选择最为合适。" }
    }

    /**
     * 生成结果数据类
     */
    data class StoryGenerationResult(
        val storySegment: String,
        val options: List<StoryOption>,
        val worldState: WorldState,
        val confidenceScore: Float
    )

    /**
     * 世界选择结果数据类
     */
    data class WorldSelectionResult(
        val recommendation: String,
        val worldOptions: List<WorldOption>,
        val reasoning: String
    )

    /**
     * 世界选项数据类
     */
    data class WorldOption(
        val worldId: String,
        val name: String,
        val description: String,
        val requiredAttributes: Map<String, Int>
    )
}

/**
 * 世界状态管理器接口
 */
interface WorldStateManager {

    /**
     * 获取玩家进度信息
     */
    suspend fun getPlayerProgress(playerId: String): PlayerProgress

    /**
     * 获取可用的世界列表
     */
    suspend fun getAvailableWorlds(playerId: String): List<String>

    /**
     * 更新世界状态
     */
    suspend fun updateWorldState(
        playerId: String,
        worldId: String,
        response: NarrativeResponse,
        context: NarrativeContext
    ): WorldState

    /**
     * 解锁新世界
     */
    suspend fun unlockWorld(playerId: String, worldId: String)

    /**
     * 记录关键事件
     */
    suspend fun recordKeyEvent(playerId: String, eventId: String, eventData: Map<String, Any>)
}

/**
 * 叙事上下文管理器接口
 */
interface NarrativeContextManager {

    /**
     * 构建叙事上下文
     */
    suspend fun buildNarrativeContext(
        playerId: String,
        worldId: String,
        playerChoice: String? = null
    ): NarrativeContext

    /**
     * 保存上下文快照
     */
    suspend fun saveContextSnapshot(
        playerId: String,
        worldId: String,
        context: NarrativeContext,
        response: NarrativeResponse
    )

    /**
     * 获取相关上下文
     */
    fun getRelevantContexts(
        playerId: String,
        worldId: String,
        contextType: String
    ): Flow<List<NarrativeContext>>

    /**
     * 清理过期上下文
     */
    suspend fun cleanupExpiredContexts(playerId: String)
}

/**
 * 玩家进度数据类
 */
data class PlayerProgress(
    val playerId: String,
    val playerAttributes: PlayerAttributes,
    val completedWorlds: List<String>,
    val activeQuests: List<String>,
    val inventoryItems: List<String>,
    val factionRelations: Map<String, String>,
    val acquiredItems: List<String>,
    val keyDecisions: List<String>,
    val currentWorld: String? = null
)

File: frontend/src/services/narrativeEngine.ts
import { 
  NarrativeContext, 
  NarrativeResponse, 
  StoryGenerationResult, 
  WorldSelectionResult, 
  WorldOption,
  PlayerProgress,
  WorldState,
  PlayerAttributes,
  StoryOption
} from '@/types'
import AIClient, { getDefaultAIClient } from './aiClient'

interface WorldStateManager {
  getPlayerProgress(playerId: string): Promise<PlayerProgress>
  getAvailableWorlds(playerId: string): Promise<string[]>
  updateWorldState(
    playerId: string,
    worldId: string,
    response: NarrativeResponse,
    context: NarrativeContext
  ): Promise<WorldState>
  unlockWorld(playerId: string, worldId: string): Promise<void>
  recordKeyEvent(playerId: string, eventId: string, eventData: Record<string, any>): Promise<void>
}

interface NarrativeContextManager {
  buildNarrativeContext(
    playerId: string,
    worldId: string,
    playerChoice?: string
  ): Promise<NarrativeContext>
  saveContextSnapshot(
    playerId: string,
    worldId: string,
    context: NarrativeContext,
    response: NarrativeResponse
  ): Promise<void>
  cleanupExpiredContexts(playerId: string): Promise<void>
}

class NarrativeEngine {
  private aiClient: AIClient
  private worldStateManager: WorldStateManager
  private contextManager: NarrativeContextManager

  constructor(
    aiClient: AIClient,
    worldStateManager: WorldStateManager,
    contextManager: NarrativeContextManager
  ) {
    this.aiClient = aiClient
    this.worldStateManager = worldStateManager
    this.contextManager = contextManager
  }

  /**
   * 生成新的故事片段
   */
  async generateStory(
    playerId: string,
    worldId: string,
    playerChoice?: string
  ): Promise<StoryGenerationResult> {
    try {
      // 1. 构建叙事上下文
      const context = await this.contextManager.buildNarrativeContext(
        playerId,
        worldId,
        playerChoice
      )

      // 2. 生成AI响应
      const rawResponse = await this.aiClient.generateNarrative(context)

      // 3. 更新世界状态
      const updatedState = await this.worldStateManager.updateWorldState(
        playerId,
        worldId,
        rawResponse,
        context
      )

      // 4. 保存上下文快照
      await this.contextManager.saveContextSnapshot(
        playerId,
        worldId,
        context,
        rawResponse
      )

      return {
        storySegment: rawResponse.storySegment,
        options: rawResponse.generatedOptions,
        worldState: updatedState,
        confidenceScore: rawResponse.confidenceScore
      }
    } catch (error) {
      console.error('故事生成失败:', error)
      throw new Error('故事生成失败，请稍后重试')
    }
  }

  /**
   * 构建世界选择上下文
   */
  async generateWorldSelection(playerId: string): Promise<WorldSelectionResult> {
    try {
      const playerProgress = await this.worldStateManager.getPlayerProgress(playerId)
      const availableWorlds = await this.worldStateManager.getAvailableWorlds(playerId)

      // 构建世界选择上下文
      const context: NarrativeContext = {
        systemInstruction: '你是一位世界引导者，需要根据玩家的成长和经历，推荐最适合的下一个故事世界。每个世界都有独特的风格和挑战。',
        worldState: {
          currentWorld: 'nexus', // 叙事行者的大本营
          keyCharacters: {},
          unlockedLocations: ['万象之根', '叙事长廊'],
          corePuzzles: {},
          storyThreads: []
        },
        recentStory: this.buildPlayerHistorySummary(playerProgress),
        playerAttributes: playerProgress.playerAttributes,
        availableWorlds,
        activeQuests: playerProgress.activeQuests,
        inventoryItems: playerProgress.inventoryItems,
        factionRelations: playerProgress.factionRelations
      }

      const response = await this.aiClient.generateNarrative(context)
      const filteredResponse = this.filterResponse(response)

      return {
        recommendation: filteredResponse.storySegment,
        worldOptions: this.extractWorldOptions(filteredResponse.generatedOptions),
        reasoning: this.extractWorldReasoning(filteredResponse.storySegment)
      }
    } catch (error) {
      console.error('世界选择生成失败:', error)
      throw new Error('世界选择生成失败，请稍后重试')
    }
  }

  /**
   * 过滤AI响应
   */
  private filterResponse(response: NarrativeResponse): NarrativeResponse {
    // 这里可以添加额外的业务逻辑过滤
    return response
  }

  /**
   * 构建玩家历史摘要
   */
  private buildPlayerHistorySummary(progress: PlayerProgress): string {
    const sb: string[] = []
    sb.push('玩家叙事行者的历史摘要：\n')

    if (progress.completedWorlds.length > 0) {
      sb.push(`已完成的世界：${progress.completedWorlds.join(', ')}\n`)
    }

    if (progress.acquiredItems.length > 0) {
      sb.push(`获得的独特物品：${progress.acquiredItems.slice(0, 5).join(', ')}\n`)
    }

    if (progress.keyDecisions.length > 0) {
      sb.push(`关键抉择：${progress.keyDecisions.slice(0, 3).join('； ')}\n`)
    }

    sb.push(`当前属性：洞察力${progress.playerAttributes.insight}，说服力${progress.playerAttributes.persuasion}，魄力${progress.playerAttributes.魄力}`)

    return sb.join('')
  }

  /**
   * 从选项中提取世界选择
   */
  private extractWorldOptions(options: StoryOption[]): WorldOption[] {
    return options.map(option => ({
      worldId: option.optionId,
      name: option.text,
      description: option.description,
      requiredAttributes: option.requiredAttributes || {}
    }))
  }

  /**
   * 从故事中提取世界推荐理由
   */
  private extractWorldReasoning(story: string): string {
    // 简单实现：提取包含"因为"、"所以"、"适合"、"推荐"等关键词的句子
    const lines = story.split('\n').filter(line => 
      line.includes('因为') || 
      line.includes('所以') || 
      line.includes('适合') || 
      line.includes('推荐')
    )
    
    return lines.join(' ').trim() || '基于你的经历和能力，这个选择最为合适。'
  }

  /**
   * 获取AI客户端
   */
  getAIClient(): AIClient {
    return this.aiClient
  }

  /**
   * 更新AI客户端配置
   */
  updateAIClientConfig(config: Partial<NarrativeResponse>) {
    // 这里可以添加配置更新逻辑
  }
}

// 默认的世界状态管理器实现
class DefaultWorldStateManager implements WorldStateManager {
  async getPlayerProgress(playerId: string): Promise<PlayerProgress> {
    // 从localStorage获取玩家进度
    const saved = localStorage.getItem(`player_progress_${playerId}`)
    if (saved) {
      return JSON.parse(saved)
    }

    // 返回默认进度
    return {
      playerId,
      playerAttributes: { insight: 10, persuasion: 10, 魄力: 10 },
      completedWorlds: [],
      activeQuests: [],
      inventoryItems: [],
      factionRelations: {},
      acquiredItems: [],
      keyDecisions: [],
      currentWorld: undefined
    }
  }

  async getAvailableWorlds(playerId: string): Promise<string[]> {
    // 基于玩家进度返回可用世界
    const progress = await this.getPlayerProgress(playerId)
    const baseWorlds = ['赛博蓬莱', '法典迷城', '衰败王座']
    
    // 可以根据完成的世界解锁新世界
    return baseWorlds
  }

  async updateWorldState(
    playerId: string,
    worldId: string,
    response: NarrativeResponse,
    context: NarrativeContext
  ): Promise<WorldState> {
    // 更新世界状态逻辑
    const newState = response.newWorldState || context.worldState
    
    // 保存进度
    const progress = await this.getPlayerProgress(playerId)
    progress.currentWorld = worldId
    
    localStorage.setItem(`player_progress_${playerId}`, JSON.stringify(progress))
    localStorage.setItem(`world_state_${playerId}_${worldId}`, JSON.stringify(newState))

    return newState
  }

  async unlockWorld(playerId: string, worldId: string): Promise<void> {
    const progress = await this.getPlayerProgress(playerId)
    if (!progress.completedWorlds.includes(worldId)) {
      progress.completedWorlds.push(worldId)
      localStorage.setItem(`player_progress_${playerId}`, JSON.stringify(progress))
    }
  }

  async recordKeyEvent(playerId: string, eventId: string, eventData: Record<string, any>): Promise<void> {
    const progress = await this.getPlayerProgress(playerId)
    progress.keyDecisions.push(`${eventId}: ${JSON.stringify(eventData)}`)
    localStorage.setItem(`player_progress_${playerId}`, JSON.stringify(progress))
  }
}

// 默认的上下文管理器实现
class DefaultNarrativeContextManager implements NarrativeContextManager {
  async buildNarrativeContext(
    playerId: string,
    worldId: string,
    playerChoice?: string
  ): Promise<NarrativeContext> {
    const progress = await this.getPlayerProgress(playerId)
    const savedState = localStorage.getItem(`world_state_${playerId}_${worldId}`)
    const worldState: WorldState = savedState ? JSON.parse(savedState) : {
      currentWorld: worldId,
      keyCharacters: {},
      unlockedLocations: [],
      corePuzzles: {},
      storyThreads: []
    }

    // 构建系统指令
    const systemInstruction = this.buildSystemInstruction(worldId)

    return {
      systemInstruction,
      worldState,
      recentStory: this.buildRecentStory(playerId, worldId),
      playerAttributes: progress.playerAttributes,
      availableWorlds: await this.getAvailableWorlds(playerId),
      activeQuests: progress.activeQuests,
      inventoryItems: progress.inventoryItems,
      factionRelations: progress.factionRelations
    }
  }

  async saveContextSnapshot(
    playerId: string,
    worldId: string,
    context: NarrativeContext,
    response: NarrativeResponse
  ): Promise<void> {
    const snapshot = {
      context,
      response,
      timestamp: Date.now()
    }
    
    // 保存最近的上下文快照
    const snapshots = JSON.parse(localStorage.getItem(`context_snapshots_${playerId}_${worldId}`) || '[]')
    snapshots.push(snapshot)
    
    // 只保留最近的10个快照
    if (snapshots.length > 10) {
      snapshots.shift()
    }
    
    localStorage.setItem(`context_snapshots_${playerId}_${worldId}`, JSON.stringify(snapshots))
  }

  async cleanupExpiredContexts(playerId: string): Promise<void> {
    // 清理过期的上下文快照
    const threeDaysAgo = Date.now() - 3 * 24 * 60 * 60 * 1000
    
    for (const key of Object.keys(localStorage)) {
      if (key.startsWith(`context_snapshots_${playerId}_`)) {
        const snapshots = JSON.parse(localStorage.getItem(key) || '[]')
        const filtered = snapshots.filter((snapshot: any) => snapshot.timestamp > threeDaysAgo)
        
        if (filtered.length === 0) {
          localStorage.removeItem(key)
        } else {
          localStorage.setItem(key, JSON.stringify(filtered))
        }
      }
    }
  }

  private async getPlayerProgress(playerId: string): Promise<any> {
    const saved = localStorage.getItem(`player_progress_${playerId}`)
    return saved ? JSON.parse(saved) : {
      playerId,
      playerAttributes: { insight: 10, persuasion: 10, 魄力: 10 },
      completedWorlds: [],
      activeQuests: [],
      inventoryItems: [],
      factionRelations: {},
      acquiredItems: [],
      keyDecisions: []
    }
  }

  private async getAvailableWorlds(playerId: string): Promise<string[]> {
    return ['赛博蓬莱', '法典迷城', '衰败王座']
  }

  private buildSystemInstruction(worldId: string): string {
    const instructions: Record<string, string> = {
      '赛博蓬莱': '你将是《赛博蓬莱》世界的叙述者。文风需融合赛博朋克的冰冷与东方古典的优雅。',
      '法典迷城': '你将是《法典迷城》世界的叙述者。文风需融合法律条文的严谨与侠盗的豪迈。',
      '衰败王座': '你将是《衰败王座》世界的叙述者。文风需融合神迹的神秘与衰败的苍凉。'
    }
    
    return instructions[worldId] || '你是一个故事世界的叙述者，请根据世界特点调整文风。'
  }

  private buildRecentStory(playerId: string, worldId: string): string {
    const snapshots = JSON.parse(localStorage.getItem(`context_snapshots_${playerId}_${worldId}`) || '[]')
    
    if (snapshots.length === 0) {
      return '故事刚开始，一切皆有可能。'
    }
    
    // 返回最近的3个故事片段
    const recent = snapshots.slice(-3).map((snapshot: any) => snapshot.response.storySegment)
    return recent.join('\n\n')
  }
}

// 创建默认实例
export function createNarrativeEngine(): NarrativeEngine {
  const aiClient = getDefaultAIClient()
  const worldStateManager = new DefaultWorldStateManager()
  const contextManager = new DefaultNarrativeContextManager()
  
  return new NarrativeEngine(aiClient, worldStateManager, contextManager)
}

export default NarrativeEngine
export { createNarrativeEngine }